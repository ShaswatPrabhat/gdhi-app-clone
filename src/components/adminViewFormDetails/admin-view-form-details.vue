<script>
import Vue from "vue";

import AdminTable from "../displayTable/admin-table.vue";
import axios from "axios";
import isEmpty from "lodash/isEmpty";

export default Vue.extend({
  name: "AdminViewFormDetails",
  components: { AdminTable },
  data() {
    return {
      selectedTab: 0,
      tableColumns: [],
      tableRows: [],
      allData: {},
      action: "",
      noRecordsMessage: "",
      error: "",
      tabs: [
        { id: 0, name: "Awaiting Submission" },
        { id: 1, name: "Review Pending" },
        { id: 2, name: "Live Data" },
      ],
    };
  },
  mounted() {
    this.loadAdminViewFormDetails();
  },
  computed: {
    currentYear() {
      return this.allData.currentYear;
    },
  },

  methods: {
    updateSelected(tab) {
      this.getTabData(tab);
      this.selectedTab = tab.id;
    },
    loadAdminViewFormDetails() {
      axios
        .get("/api/countries/country_status_summaries")
        .then((response) => {
          this.allData = response.data;
          this.updateSelected(this.tabs[0]);
        })
        .catch((e) => {
          this.error = e.response.message;
        });
    },
    actionHandler(action, countryUUID) {
      if (action === "Review") {
        this.openUrl(
          location.origin +
            "/admin/health_indicator_questionnaire/" +
            countryUUID +
            "/review/" +
            this.currentYear
        );
      } else if (action === "View Live Data") {
        this.openUrl(
          location.origin +
            "/admin/health_indicator_questionnaire/" +
            countryUUID +
            "/viewPublished/" +
            this.currentYear
        );
      } else if (action === "Edit Live Data") {
        console.log("EDIT LIVE DATA");
        this.openUrl(
          location.origin +
            "/admin/health_indicator_questionnaire/" +
            countryUUID +
            "/editPublished/" +
            this.currentYear
        );
      }
    },
    openUrl(url) {
      window.open(url);
    },
    wrapperOnTableRows(rows) {
      rows.forEach(function (row) {
        row.url =
          location.origin +
          "/health_indicator_questionnaire/" +
          row.countryUUID;
      });
    },
    getTabData(tab) {
      this.tableColumns = [];
      this.tableRows = [];
      this.noRecordsMessage = "";
      if (isEmpty(this.allData)) {
        this.noRecordsMessage = "No Records Found";
        return;
      }
      switch (tab) {
        case this.tabs[0]:
          this.tableColumns = [
            { propName: "countryName", displayName: "Country" },
            { propName: "status", displayName: "Status" },
          ];

          this.tableRows = [];
          if (this.allData.NEW !== undefined)
            this.tableRows = [...this.tableRows, ...this.allData.NEW];
          if (this.allData.DRAFT !== undefined)
            this.tableRows = [...this.tableRows, ...this.allData.DRAFT];
          this.wrapperOnTableRows(this.tableRows);
          this.action = "View Data";
          break;
        case this.tabs[1]:
          this.tableColumns = [
            { propName: "countryName", displayName: "Country" },
            { propName: "contactName", displayName: "Country Contact Name" },
            { propName: "contactEmail", displayName: "Country Contact Email" },
          ];
          this.action = "Review";
          this.tableRows = [];
          if (this.allData.REVIEW_PENDING !== undefined) {
            this.tableRows = this.allData.REVIEW_PENDING;
          }
          break;
        case this.tabs[2]:
          this.tableColumns = [
            { propName: "countryName", displayName: "Country" },
            { propName: "contactName", displayName: "Country Contact Name" },
            { propName: "contactEmail", displayName: "Country Contact Email" },
          ];
          this.action = "View Live Data";
          this.tableRows = [];
          if (this.allData.PUBLISHED !== undefined) {
            this.tableRows = this.allData.PUBLISHED;
          }
          break;
      }
      if (this.tableRows.length <= 0) {
        this.noRecordsMessage = "No Records Found";
      }
    },
  },
});
</script>

<template>
  <div>
    <div class="header-bold">Form Details for {{ currentYear }} year</div>
    <div class="tabs">
      <ul class="tablist">
        <li
          v-for="tab in tabs"
          v-on:click="updateSelected(tab)"
          :class="{ current: selectedTab == tab.id }"
        >
          {{ tab.name }}
        </li>
      </ul>
      <div
        class="tab-content"
        v-for="tab in tabs"
        v-show="selectedTab == tab.id"
      >
        <AdminTable
          :noRecordsMessage="noRecordsMessage"
          :columns="tableColumns"
          :rows="tableRows"
          :action="action"
          :actionHandler="actionHandler"
        ></AdminTable>
      </div>
    </div>
  </div>
</template>
