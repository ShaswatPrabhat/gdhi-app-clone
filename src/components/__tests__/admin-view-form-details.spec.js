import { mount } from "@vue/test-utils";
import { describe, expect, it, vi } from "vitest";
import adminViewFormDetails from "../adminViewFormDetails/admin-view-form-details.vue";
import VueRouter from "vue-router";
import { i18n } from "../../plugins/i18n";
import axios from "axios";
import flushPromises from "flush-promises";

const axiosGetSpy = vi.spyOn(axios, "get");

describe("AdminViewFormDetails", () => {
  let component;
  const router = new VueRouter();

  let responseData = {
    currentYear: "2023",
    NEW: [
      {
        countryName: "India",
        countryUUID: "1b9f0f33-9c69-4bfc-8730-8bcbb5c06d88",
        status: "NEW",
        contactName: "contactName",
        contactEmail: "email",
      },
      {
        countryName: "Australia",
        countryUUID: "5a1f8e8c-7ebd-499a-b6b0-015828695796",
        status: "NEW",
        contactName: "contactName",
        contactEmail: "email",
      },
    ],
    PUBLISHED: [
      {
        countryName: "Australia",
        countryUUID: "5a1f8e8c-7ebd-499a-b6b0-015828695796",
        status: "PUBLISHED",
        contactName: "contactName",
        contactEmail: "email",
      },
    ],
    DRAFT: [
      {
        countryName: "Australia",
        countryUUID: "5a1f8e8c-7ebd-499a-b6b0-015828695796",
        status: "DRAFT",
        contactName: "contactName",
        contactEmail: "email",
      },
    ],
    REVIEW_PENDING: [
      {
        countryName: "Australia",
        countryUUID: "5a1f8e8c-7ebd-499a-b6b0-015828695796",
        status: "REVIEW_PENDING",
        contactName: "contactName",
        contactEmail: "email",
      },
    ],
  };
  it("should get admin view form details data", async () => {
    axiosGetSpy.mockResolvedValue({ data: responseData });

    component = mount(adminViewFormDetails, {
      data: () => ({
        tabs: [
          { id: 0, name: "Awaiting Submission" },
          { id: 1, name: "Review Pending" },
          { id: 2, name: "Live Data" },
        ],
      }),
      router,
    });
    await flushPromises();

    let updateSelected = vi.fn();
    component.vm.updateSelected = updateSelected;
    component.vm.loadAdminViewFormDetails();

    await flushPromises();

    updateSelected.mockReturnValue({
      id: 0,
      name: "Awaiting Submission",
    });

    expect(component.vm.allData).to.equal(responseData);
    expect(component.vm.$el).toMatchSnapshot();
  });

  it("should call getTabData when updateSelected is called", async () => {
    axiosGetSpy.mockResolvedValue({ data: responseData });

    component = mount(adminViewFormDetails, {
      data: () => ({
        tabs: [
          { id: 0, name: "Awaiting Submission" },
          { id: 1, name: "Review Pending" },
          { id: 2, name: "Live Data" },
        ],
      }),
      router,
    });
    await flushPromises();

    let getTabData = vi.fn();
    component.vm.getTabData = getTabData;
    component.vm.updateSelected({ id: 0, name: "Awaiting Submission" });
    getTabData.mockReturnValue({
      id: 0,
      name: "Awaiting Submission",
    });

    expect(component.vm.selectedTab).to.equal(0);
  });

  it("should set error value when the API call is failed", async () => {
    const errResp = {
      response: { message: "problem" },
    };
    axiosGetSpy.mockRejectedValue(errResp);
    component = mount(adminViewFormDetails, {
      data: () => ({
        tabs: [
          { id: 0, name: "Awaiting Submission" },
          { id: 1, name: "Review Pending" },
          { id: 2, name: "Live Data" },
        ],
      }),
      router,
    });
    await flushPromises();

    expect(component.vm.error).to.equal(errResp.response.message);
  });

  it("should call openUrl when actionHandler is invoked", async () => {
    axiosGetSpy.mockResolvedValue({ data: responseData });

    component = mount(adminViewFormDetails, {
      data: () => ({
        tabs: [
          { id: 0, name: "Awaiting Submission" },
          { id: 1, name: "Review Pending" },
          { id: 2, name: "Live Data" },
        ],
      }),
      router,
      i18n,
    });
    await flushPromises();
    let openUrl = vi.fn();
    component.vm.openUrl = openUrl;
    openUrl.mockReturnValueOnce(
      location.origin + "/admin/health_indicator_questionnaire/some-uuid/review"
    );

    component.vm.actionHandler("Review", "some-uuid");
    expect(openUrl.mock.calls.length).to.equal(1);
    openUrl.mockReturnValueOnce(
      location.origin + "/admin/health_indicator_questionnaire/some-uuid/review"
    );

    component.vm.actionHandler("View Live Data", "some-uuid");
    expect(openUrl.mock.calls.length).to.equal(2);

    component.vm.actionHandler("Other Text", "some-uuid");
    expect(openUrl.mock.calls.length).to.equal(2);
  });

  it("should call openUrl when editHandler is invoked", async () => {
    axiosGetSpy.mockResolvedValue({ data: responseData });

    component = mount(adminViewFormDetails, {
      data: () => ({
        tabs: [
          { id: 0, name: "Awaiting Submission" },
          { id: 1, name: "Review Pending" },
          { id: 2, name: "Live Data" },
        ],
      }),
      router,
      i18n,
    });
    await flushPromises();
    let openUrl = vi.fn();
    component.vm.openUrl = openUrl;
    openUrl.mockReturnValueOnce(
      location.origin +
        "/admin/health_indicator_questionnaire/some-uuid/editPublished"
    );

    component.vm.editHandler("some-uuid");
    expect(openUrl.mock.calls.length).to.equal(1);
  });

  it("should populate the table rows when getTabData is called ", async () => {
    axiosGetSpy.mockResolvedValue({ data: responseData });

    component = mount(adminViewFormDetails, {
      data: () => ({
        tabs: [
          { id: 0, name: "Awaiting Submission" },
          { id: 1, name: "Review Pending" },
          { id: 2, name: "Live Data" },
        ],
      }),
      router,
    });
    await flushPromises();

    component.vm.allData = responseData;
    component.vm.getTabData(component.vm.tabs[0]);
    expect(component.vm.tableRows).to.deep.equal([
      ...responseData.NEW,
      ...responseData.DRAFT,
    ]);
    expect(component.vm.noRecordsMessage).to.equal("");
    component.vm.getTabData(component.vm.tabs[1]);
    expect(component.vm.tableRows).to.deep.equal(responseData.REVIEW_PENDING);
    expect(component.vm.editLiveData).toBe(false);
    expect(component.vm.noRecordsMessage).to.equal("");
    component.vm.getTabData(component.vm.tabs[2]);
    expect(component.vm.tableRows).to.deep.equal(responseData.PUBLISHED);
    expect(component.vm.noRecordsMessage).to.equal("");
    expect(component.vm.editLiveData).toBe(true);
  });

  it("should return empty tablerows when the value is undefined", async () => {
    axiosGetSpy.mockResolvedValue({ data: responseData });

    component = mount(adminViewFormDetails, {
      data: () => ({
        tabs: [
          { id: 0, name: "Awaiting Submission" },
          { id: 1, name: "Review Pending" },
          { id: 2, name: "Live Data" },
        ],
      }),
      router,
    });
    await flushPromises();

    let updatedResponse = { ...responseData };
    updatedResponse.NEW = undefined;
    component.vm.allData = updatedResponse;
    component.vm.getTabData(component.vm.tabs[0]);
    expect(component.vm.tableRows).to.deep.equal([...responseData.DRAFT]);

    updatedResponse = { ...responseData };
    updatedResponse.DRAFT = undefined;
    component.vm.allData = updatedResponse;
    component.vm.getTabData(component.vm.tabs[0]);
    expect(component.vm.tableRows).to.deep.equal([...responseData.NEW]);

    updatedResponse = { ...responseData };
    updatedResponse.NEW = undefined;
    updatedResponse.DRAFT = undefined;
    component.vm.allData = updatedResponse;
    component.vm.getTabData(component.vm.tabs[0]);
    expect(component.vm.tableRows).to.deep.equal([]);
    expect(component.vm.noRecordsMessage).to.equal("No Records Found");

    updatedResponse = { ...responseData };
    updatedResponse.REVIEW_PENDING = undefined;
    component.vm.allData = updatedResponse;
    component.vm.getTabData(component.vm.tabs[1]);
    expect(component.vm.tableRows).to.deep.equal([]);
    expect(component.vm.noRecordsMessage).to.equal("No Records Found");

    updatedResponse = { ...responseData };
    updatedResponse.PUBLISHED = undefined;
    component.vm.allData = updatedResponse;
    component.vm.getTabData(component.vm.tabs[2]);
    expect(component.vm.tableRows).to.deep.equal([]);
    expect(component.vm.noRecordsMessage).to.equal("No Records Found");

    component.vm.allData = {};
    component.vm.getTabData(component.vm.tabs[2]);
    expect(component.vm.tableRows).to.deep.equal([]);
    expect(component.vm.noRecordsMessage).to.equal("No Records Found");
  });

  it("should set the tablerows to [] when getTab data is called for indices greater than 2", async () => {
    axiosGetSpy.mockResolvedValue({ data: responseData });

    component = mount(adminViewFormDetails, {
      data: () => ({
        tabs: [
          { id: 0, name: "Awaiting Submission" },
          { id: 1, name: "Review Pending" },
          { id: 2, name: "Live Data" },
        ],
      }),
      router,
    });
    await flushPromises();

    component.vm.allData = responseData;
    component.vm.getTabData(component.vm.tabs[3]);
    expect(component.vm.tableRows).to.deep.equal([]);
    expect(component.vm.tableColumns).to.deep.equal([]);
  });
});
