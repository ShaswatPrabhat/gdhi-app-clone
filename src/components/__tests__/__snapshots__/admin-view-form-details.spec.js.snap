// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AdminViewFormDetails > should get admin view form details data 1`] = `
<div>
  <div
    class="header-bold"
  >
    Form Details for 2023 year
  </div>
  <div
    class="tabs"
  >
    <ul
      class="tablist"
    >
      <li
        class="current"
      >
         Awaiting Submission 
      </li>
      <li
        class=""
      >
         Review Pending 
      </li>
      <li
        class=""
      >
         Live Data 
      </li>
    </ul>
    <div
      class="tab-content"
    >
      <div>
        <!---->
        <table
          id="fifthTable"
        >
          <thead>
            <tr>
              <th>
                Country
              </th>
              <th>
                Status
              </th>
              <th>
                <span>
                  URL
                </span>
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                 India 
              </td>
              <td>
                 NEW 
              </td>
              <td>
                http://localhost:3000/health_indicator_questionnaire/1b9f0f33-9c69-4bfc-8730-8bcbb5c06d88
              </td>
            </tr>
            <tr>
              <td>
                 Australia 
              </td>
              <td>
                 NEW 
              </td>
              <td>
                http://localhost:3000/health_indicator_questionnaire/5a1f8e8c-7ebd-499a-b6b0-015828695796
              </td>
            </tr>
            <tr>
              <td>
                 Australia 
              </td>
              <td>
                 DRAFT 
              </td>
              <td>
                http://localhost:3000/health_indicator_questionnaire/5a1f8e8c-7ebd-499a-b6b0-015828695796
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      class="tab-content"
      style="display: none;"
    >
      <div>
        <!---->
        <table
          id="fifthTable"
        >
          <thead>
            <tr>
              <th>
                Country
              </th>
              <th>
                Status
              </th>
              <th>
                <span>
                  URL
                </span>
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                 India 
              </td>
              <td>
                 NEW 
              </td>
              <td>
                http://localhost:3000/health_indicator_questionnaire/1b9f0f33-9c69-4bfc-8730-8bcbb5c06d88
              </td>
            </tr>
            <tr>
              <td>
                 Australia 
              </td>
              <td>
                 NEW 
              </td>
              <td>
                http://localhost:3000/health_indicator_questionnaire/5a1f8e8c-7ebd-499a-b6b0-015828695796
              </td>
            </tr>
            <tr>
              <td>
                 Australia 
              </td>
              <td>
                 DRAFT 
              </td>
              <td>
                http://localhost:3000/health_indicator_questionnaire/5a1f8e8c-7ebd-499a-b6b0-015828695796
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      class="tab-content"
      style="display: none;"
    >
      <div>
        <!---->
        <table
          id="fifthTable"
        >
          <thead>
            <tr>
              <th>
                Country
              </th>
              <th>
                Status
              </th>
              <th>
                <span>
                  URL
                </span>
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                 India 
              </td>
              <td>
                 NEW 
              </td>
              <td>
                http://localhost:3000/health_indicator_questionnaire/1b9f0f33-9c69-4bfc-8730-8bcbb5c06d88
              </td>
            </tr>
            <tr>
              <td>
                 Australia 
              </td>
              <td>
                 NEW 
              </td>
              <td>
                http://localhost:3000/health_indicator_questionnaire/5a1f8e8c-7ebd-499a-b6b0-015828695796
              </td>
            </tr>
            <tr>
              <td>
                 Australia 
              </td>
              <td>
                 DRAFT 
              </td>
              <td>
                http://localhost:3000/health_indicator_questionnaire/5a1f8e8c-7ebd-499a-b6b0-015828695796
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
